docs/source/developer/ui_components.rst

```diff
@@ -0,0 +1,337 @@
+     marginLeft: '2px'
+   });
+
+where ``elem`` is any ``HTMLElement`` with a ``div`` or ``span`` tag. As
+shown in the above example, the icon can be styled by passing CSS
+parameters into ``.element(...)``. Any valid CSS parameter can be used
+(one catch: snake case params do have to be converted to camel case:
+instead of ``foo-bar: '8px'``, you’d need to use ``fooBar: '8px'``.
+
+How to render an icon as a React component
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Icons can also be rendered using React. The ``icon.react`` parameter
+holds a standard React component that will display the icon on render.
+Like any React component, ``icon.react`` can be used in various ways.
+
+For example, here is how you would add the Jupyter icon to the render
+tree of another React component:
+
+.. code:: jsx
+
+   public render() {
+     return (
+       <div className="outer">
+         <div className="inner">
+           <jupyterIcon.react tag="span" right="7px" top="5px" />
+           "and here's a text node"
+         </div>
+       </div>
+     );
+   }
+
+Alternatively, you can just render the icon directly into any existing
+DOM node ``elem`` by using the ``ReactDOM`` module:
+
+.. code:: typescript
+
+   ReactDOM.render(jupyterIcon.react, elem);
+
+If do you use ``ReactDOM`` to render, and if the ``elem`` node is ever
+removed from the DOM, you’ll first need to clean it up:
+
+.. code:: typescript
+
+   ReactDOM.unmountComponentAtNode(elem);
+
+This cleanup step is not a special property of ``LabIcon``, but is
+instead needed for any React component that is rendered directly at the
+top level by ``ReactDOM``: failure to call ``unmountComponentAtNode``
+can result in a `memory
+leak <https://stackoverflow.com/a/48198011/425458>`__.
+
+How to create your own custom ``LabIcon``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can create your own custom icon by constructing a new instance of
+``LabIcon``:
+
+.. code:: typescript
+
+   export const fooIcon = new LabIcon({
+     name: 'barpkg:foo',
+     svgstr: '<svg>...</svg>'
+   });
+
+where ``name`` should be of the form “your-pkg:icon-name”, and
+``svgstr`` is the raw contents of your icon’s svg file.
+
+How to create a new ``LabIcon`` from an external svg file
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Although you can copy-and-paste an svg directly into the ``LabIcon``
+constructor, the best practice is to keep the svg for each of your icons
+in its own separate svg file. You will need to have an ``svg.d.ts`` file
+at the root of your project’s ``src`` directory:
+
+.. code:: typescript
+
+   // svg.d.ts
+
+   declare module '*.svg' {
+     const value: string;
+     export default value;
+   }
+
+You can then ``import`` the contents of an svg file:
+
+.. code:: typescript
+
+   import fooSvgstr from 'path-to-your/foo.svg';
+
+   export const fooIcon = new LabIcon({
+     name: 'barpkg:foo',
+     svgstr: fooSvgstr
+   });
+
+Sync icon color to JupyterLab theme
```

Comment:
I am not too keen about having duplicate text in three places checked into VCS, because it seems likely we might update it in one place and forget the other two. Any ideas on how to make this have a single source of truth?

Reply:
> Any ideas on how to make this have a single source of truth?

Nope, but I'll take suggestions :)

More seriously, I do have some notions about how to improve the `docs:init` build here, but they're all waaay out of scope for what's supposed to be a PR updating the LabIcon docs. I am actively working on improving the API docs, and may be able to fold this issue into that work as well.

For now, we treat the sources in `ui-components/docs` as the source of truth. Running `jlpm docs:init` (while in `ui-components`) will automatically propagate any changes to the README and the dev docs. If anyone does end up commiting changes directly to the wrong copy, I'll deal with manually resolving the conflicts

Reply:
I've now added a warning comment to the top of the autogenerated files:

```html
<!--
THIS FILE IS AUTOGENERATED, DO NOT EDIT

Instead, make changes to docs sources in `packages/ui-components/docs`,
then run "jlpm docs:init" to refresh the built docs
-->
```

Reply:
> Running jlpm docs:init (while in ui-components) will automatically propagate any changes to the README and the dev docs. If anyone does end up commiting changes directly to the wrong copy, I'll deal with manually resolving the conflicts

For the README maybe we could just link to that file instead of copying? And maybe for the docs we could have it do that live generation when generating the docs instead of storing it in VCS?

What do you think?

Reply:
> For the README maybe we could just link to that file instead of copying? And maybe for the docs we could have it do that live generation when generating the docs instead of storing it in VCS?

As I said in an earlier comment, that is exactly how I had things set up earlier in this PR at b921254. The problem was that linking from `ui-components/README.md` to another markdown file doesn't work in the rendered API docs.

In the future I may end up just adding a link in the README.md to the live rendered LabIcon dev docs on readthedocs. However, that will be completely broken until the new LabIcon docs have been pushed to readthedocs in the first place, so I think it's best to do that in the next PR

Reply:
basically, I don't want to roll something project wide and complicated that I'm just going to remove in a future PR. Keeping the docs checked in in 3 places is dumb, but it's also simple and renders reliably